timemachine/fe/free_energy.py

```diff
@@ -392,12 +427,64 @@ def _combine(self, ligand_values, host_values=None):
+    summed_pot = next(p.potential for p in initial_state.potentials if isinstance(p.potential, SummedPotential))
+    # TBD Figure out a better way of handling the jankyness of having to select the IxnGroup that defines the Ligand-Water
+    # Currently this just hardcodes to the first ixn group potential which is the ligand-water ixn group as returned by HostTopology.parameterize_nonbonded
+    ixn_group_idx = next(i for i, pot in enumerate(summed_pot.potentials) if isinstance(pot, NonbondedInteractionGroup))
+    assert isinstance(summed_pot.potentials[ixn_group_idx], NonbondedInteractionGroup)
+    water_params = summed_pot.params_init[ixn_group_idx]
+    assert water_params.shape[1] == 4
+    return water_params
+
+
+def get_context(initial_state: InitialState, md_params: Optional[MDParams] = None) -> Context:
+    """
+    Construct a Context that has a single SummedPotential that combines the potentials defined by the initial state
+    """
+    potentials = [bp.potential for bp in initial_state.potentials]
+    params = [bp.params for bp in initial_state.potentials]
+    potential = SummedPotential(potentials, params).to_gpu(np.float32)
+
+    # Set up context for MD using overall potential
+    bound_impl = potential.bind_params_list(params).bound_impl
+    bound_impls = [bound_impl]
     intg_impl = initial_state.integrator.impl()
     movers = []
     if initial_state.barostat:
         movers.append(initial_state.barostat.impl(bound_impls))
+    if md_params is not None and md_params.water_sampling_params is not None:
+        # Setup the water indices
+        hb_potential = next(p.potential for p in initial_state.potentials if isinstance(p.potential, HarmonicBond))
+        group_indices = get_group_indices(get_bond_list(hb_potential), len(initial_state.integrator.masses))
+
+        water_idxs = [group for group in group_indices if len(group) == 3]
+        # Handle the case where the ligand is a 3 atom system
+        if len(initial_state.ligand_idxs) == 3:
+            ligand_atom_set = set(initial_state.ligand_idxs)
+            water_idxs = [group for group in water_idxs if ligand_atom_set != set(group)]
+
+        summed_pot = next(p.potential for p in initial_state.potentials if isinstance(p.potential, SummedPotential))
+        # Select a Nonbonded Potential to get the the cutoff/beta, assumes all have same cutoff/beta.
+        nb = next(p for p in summed_pot.potentials if isinstance(p, NonbondedInteractionGroup))
+
+        water_params = get_water_params(initial_state)
+
+        water_sampler = custom_ops.TIBDExchangeMove_f32(
+            initial_state.x0.shape[0],
+            initial_state.ligand_idxs.tolist(),
+            [water_group.tolist() for water_group in water_idxs],
+            water_params,
+            initial_state.integrator.temperature,
+            nb.beta,
+            nb.cutoff,
+            md_params.water_sampling_params.radius,
+            md_params.seed,
```

Comment:
Should we add `+3` or do some sort of modification to the seed here?

Reply:
Probably? (Otherwise there's probably some minor correlation between the different RNGs)

Reply:
Generates new seed on the fly in https://github.com/proteneer/timemachine/pull/1286/commits/5d10059bb7c71dd39c1e02b19cc2945c96067897